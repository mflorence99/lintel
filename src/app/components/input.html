<ng-content select="[prefix]"></ng-content>

<input
  (input)="value = $event.target.value"
  [attr.disabled]="!enabled? true : null"
  [max]="max"
  [min]="min"
  [placeholder]="placeholder"
  [step]="step"
  [type]="type"
  [value]="value"
  #input 
  autocomplete="off"
  tabIndex=0>

<ng-content select="[suffix]"></ng-content>
