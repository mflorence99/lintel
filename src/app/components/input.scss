:host {
  align-items: center;
  background: var(--vscode-settings-dropdownBackground);
  border: 1px solid var(--vscode-settings-dropdownBorder);
  color: var(--vscode-settings-dropdownForeground);
  display: flex;
  height: 2rem;
  padding: 0.4rem;
}

:host(.focus) {
  border-color: var(--vscode-focusBorder);
}

input {
  width: 100%;
}

::ng-deep lintel-input > fa-icon {
  opacity: var(--lintel-opacity-icon);
}

::ng-deep lintel-input > fa-icon.disabled {
  opacity: var(--lintel-opacity-disabled);
}

::ng-deep lintel-input [prefix] {
  margin-right: 0.5rem;
}

::ng-deep lintel-input [suffix] {
  // NOTE: suffix is assumed to be a call to action
  cursor: pointer;
  margin-left: 0.5rem;
}
