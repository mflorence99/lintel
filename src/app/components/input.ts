import { ChangeDetectionStrategy } from '@angular/core';
import { ChangeDetectorRef } from '@angular/core';
import { Component } from '@angular/core';
import { ControlValueAccessor } from '@angular/forms';
import { ElementRef } from '@angular/core';
import { Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ViewChild } from '@angular/core';

import { forwardRef } from '@angular/core';

export type InputValueType = number | string;

/**
 * Simple input form control
 *
 * NOTE: just enough to be able to match VSCode as well as possible
 *
 * @see https://blog.thoughtram.io/angular/2016/07/27/custom-form-controls-in-angular-2.html
 *
 * Reference is old, but still helpful
 */

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => InputComponent),
      multi: true
    }
  ],
  selector: 'lintel-input',
  templateUrl: 'input.html',
  styleUrls: ['input.scss']
})
export class InputComponent implements ControlValueAccessor {
  @Input() enabled = true;

  @ViewChild('input', { static: true }) input: ElementRef;

  @Input() max: number;
  @Input() min: number;

  @Input() placeholder = '';

  @Input() step: number;

  @Input() type: 'number' | 'text' = 'text';

  @Input()
  get value(): InputValueType {
    return this._value;
  }
  set value(value: InputValueType) {
    this._value = value;
    if (this.input?.nativeElement) this.input.nativeElement.value = value;
    this.onChange?.(this.type === 'number' ? Number(value) : value);
    this.cdf.markForCheck();
  }

  private _value: InputValueType;
  private onChange: Function;

  /** ctor */
  constructor(private cdf: ChangeDetectorRef) {}

  /** @see ControlValueAccessor */
  registerOnChange(fn): void {
    this.onChange = fn;
  }

  /** @see ControlValueAccessor */
  registerOnTouched(): void {}

  /** @see ControlValueAccessor */
  writeValue(value): void {
    this.value = value;
  }
}
