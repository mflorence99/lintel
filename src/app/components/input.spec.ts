import { InputComponent } from './input';
import { TestBed } from '@angular/core/testing';

import { async } from '@angular/core/testing';
import { prepare } from './component.spec';

describe('InputComponent', () => {

  beforeEach(async(() => prepare()));

  test('Component is created', () => {
    const fixture = TestBed.createComponent(InputComponent);
    const component = fixture.componentInstance;
    expect(component).toBeTruthy();
  });

  test('registerOnChange', () => {
    const fixture = TestBed.createComponent(InputComponent);
    const component = fixture.componentInstance;
    component.registerOnChange(jest.fn());
    expect(component['onChange']).toBeTruthy();
  });

  test('registerOnTouched', () => {
    const fixture = TestBed.createComponent(InputComponent);
    const component = fixture.componentInstance;
    expect(component.registerOnTouched(null)).toBeFalsy();
  });

  test('writeValue', () => {
    const fixture = TestBed.createComponent(InputComponent);
    const component = fixture.componentInstance;
    expect(component.value).toBeUndefined();
    component.writeValue(42);
    expect(component.value).toBe(42);
  });

  test('set number value', () => {
    const fixture = TestBed.createComponent(InputComponent);
    const component = fixture.componentInstance;
    component.type = 'number';
    expect(component.value).toBeFalsy();
    component.registerOnChange(value => expect(value).toBe(42));
    component.value = 42;
    expect(component.value).toBe(42);
  });

  test('set string value', () => {
    const fixture = TestBed.createComponent(InputComponent);
    const component = fixture.componentInstance;
    component.type = 'text';
    expect(component.value).toBeFalsy();
    component.registerOnChange(value => expect(value).toBe('xxx'));
    component.value = 'xxx';
    expect(component.value).toBe('xxx');
  });

});
