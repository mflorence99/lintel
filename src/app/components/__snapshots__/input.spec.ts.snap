// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputComponent snapshot 1`] = `
<lintel-input
  __ngContext__={[Function LRootView]}
  cdf={[Function ViewRef]}
  enabled={[Function Boolean]}
  input={[Function ElementRef]}
  placeholder=""
  type={[Function String]}
>
  <input
    autocomplete="off"
    max="undefined"
    min="undefined"
    placeholder=""
    step="undefined"
    tabindex="0"
    type="text"
  />
</lintel-input>
`;
